cluster-autoscaler/cloudprovider/azure/azure_vms_pool.go

```diff
@@ -102,28 +120,260 @@ func (agentPool *VMsPool) MaxSize() int {
+	versionedAP, err := agentPool.getAgentpool()
+	if err != nil {
+		klog.Errorf("Failed to get vms pool %s, error: %s", agentPool.Name, err)
+		return armcontainerservice.AgentPool{}, err
+	}
+
+	if versionedAP.Properties != nil &&
+		versionedAP.Properties.VirtualMachinesProfile != nil &&
+		versionedAP.Properties.VirtualMachinesProfile.Scale != nil {
+		requestBody := armcontainerservice.AgentPool{
+			Properties: &armcontainerservice.ManagedClusterAgentPoolProfileProperties{
+				Type: versionedAP.Properties.Type,
+			},
+		}
+
+		// the request body must have the same mode as the original agentpool
+		// otherwise the PUT request will fail
+		if versionedAP.Properties.Mode != nil &&
+			*versionedAP.Properties.Mode == armcontainerservice.AgentPoolModeSystem {
+			systemMode := armcontainerservice.AgentPoolModeSystem
+			requestBody.Properties.Mode = &systemMode
+		}
+
+		// self-hosted CAS will be using Manual scale profile
+		if len(versionedAP.Properties.VirtualMachinesProfile.Scale.Manual) > 0 {
+			// set the count for first manual scale profile to the new target value
+			virtualMachineProfile := *versionedAP.Properties.VirtualMachinesProfile
+			versionedAP.Properties.VirtualMachinesProfile.Scale.Manual[0].Count = to.Int32Ptr(int32(count))
+			requestBody.Properties.VirtualMachinesProfile = &virtualMachineProfile
+			return requestBody, nil
+		}
+		// aks-managed CAS will be using AutoscaleProfile
+		// TODO(wenxuan): support AutoscaleProfile once it become available with the new API release
+	}
+	return armcontainerservice.AgentPool{}, fmt.Errorf("failed to build request body for scale up, agentpool doesn't have valid virtualMachinesProfile")
+}
+
+// scaleUpToCount sets node count for vms agent pool to target value through PUT AP call.
+func (agentPool *VMsPool) scaleUpToCount(count int64) error {
+	agentPool.sizeMutex.Lock()
+	defer agentPool.sizeMutex.Unlock()
+
+	updateCtx, updateCancel := getContextWithCancel()
+	defer updateCancel()
+
+	requestBody, err := agentPool.buildRequestBodyForScaleUp(count)
+	if err != nil {
+		klog.Errorf("Failed to build request body for scale up, error: %s", err)
+		return err
+	}
+
+	klog.V(1).Infof("Scaling up vms pool %s to new target count: %d", agentPool.Name, count)
+	poller, err := agentPool.manager.azClient.agentPoolClient.BeginCreateOrUpdate(
+		updateCtx,
+		agentPool.clusterResourceGroup,
+		agentPool.clusterName,
+		agentPool.Name,
+		requestBody, nil)
+
+	if err != nil {
+		klog.Errorf("Failed to update agentpool %s in cluster %s with error: %v",
+			agentPool.azureRef.Name, agentPool.clusterName, err)
+		return err
+	}
+
+	updateCtx, cancel := getContextWithTimeout(updateRequestContextTimeout)
+	defer cancel()
+
+	if _, err = poller.PollUntilDone(updateCtx, nil); err == nil {
+		// success path
+		klog.Infof("agentPoolClient.BeginCreateOrUpdate for aks cluster %s agentpool %s succeeded", agentPool.clusterName, agentPool.Name)
```

Comment:
nit: do we always want to log on success? Won't we in most cases log the results of this function ie the error? So we can infer when its successful that way? 


We don't do a good job of this now, but really we should take better care across the codebase with verbosity.

Reply:
Log removed